README
 _____ ____  __  __           _____                 _       _
|_   _|  _ \|  \/  |         | ____|_ __ ___  _   _| | __ _| |_ ___  _ __
  | | | |_) | |\/| |  _____  |  _| | '_ ` _ \| | | | |/ _` | __/ _ \| '__|
  | | |  __/| |  | | |_____| | |___| | | | | | |_| | | (_| | || (_) | |
  |_| |_|   |_|  |_|         |_____|_| |_| |_|\__,_|_|\__,_|\__\___/|_|

 TPM-Emulator v0.3 - 
 A Software-Based Trusted Platform Module (TPM) Emulator for Linux.
--------------------------------------------------------------------------

$Id$

Copyright
--------------------------------------------------------------------------
Copyright (C) 2004 Mario Strasser <mast@gmx.net> and Swiss Federal 
Institute of Technology (ETH) Zurich.
              
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


Configuration of GNU MP Library
--------------------------------------------------------------------------
Make sure that the GNU MP library is installed on your system (i.e.,
that the library file libgmp.a and the header file gmp.h exist). 
If not, you can download it from: http://www.gnu.org/software/gmp/gmp.html
or use one of the pre-compiled binaries for your distribution.

By defining the variables GMP_LIB and GMP_HEADER to the static GNU MP
library and the corresponding header file, respectively, some of the 
highly optimised GNU MP core functions (mpz_*, mpn_*) are linked against 
the kernel module. The possibility to use the adapted generic GNU MP 
sources is no longer supported.

Example:
GMP_LIB        := /usr/lib/libgmp.a
GMP_HEADER     := /usr/include/gmp.h

Installation
--------------------------------------------------------------------------
The compilation and installation process uses the build environment for 
external kernel modules of the 2.5.X Linux kernels, which must therefore
be setup properly. If you are using a pre-compiled standard kernel of 
some distribution, install the appropriate kernel-source packages and call
the following commands:

bash# cd /usr/src/linux
bash# make cloneconfig
bash# make prepare-all

The actual compilation and installation of the module is done as follows:

bash# tar -xvzf tpm_emulator-X.Y.tar.gz
bash# cd tpm_emulator-X.Y
bash# make
bash# make install


Startup
--------------------------------------------------------------------------
The startup mode of the TPM emulator (see TPM Spec. Part 1, section 7.3) 
can be defined by means of a module parameter called startup, which can 
either be set to clear, save (default) or deactivated.

Note: If the emulator is started in mode 'save' and fails to load
a previously stored TPM state it will go into fail-stop mode and has
to be reloaded. Therefore, the first time the emulator module is 
loaded, startup SHOULD be set to 'clear':

bash# modprobe tpm_emulator startup="clear"

In order to recover a TPM emulator which is in fail-stop mode, it first
has to be deactivated and then reloaded in mode 'clear':

bash# modprobe tpm_emulator startup="deactivated"
bash# rmmod tpm_emulator
bash# modprobe tpm_emulator startup="clear"

Notes
--------------------------------------------------------------------------
* The name as well as the format of the persistent-storage file has 
  changed between release 0.2 and 0.3.
* The DAA support was tested with the IBM DAA Test Suite and should work 
  now as expected. Thanks go to Roger Zimmermann for his kindly help.

Description
--------------------------------------------------------------------------
TODO: 


Contact
--------------------------------------------------------------------------
Any comments, suggestions and bug reports are welcome. Please, mention
the keyword 'TPM-Emulator' in the subject.
Mario Strasser <mast@gmx.net>



ChangeLog
2006-08-01  Heiko Stamer <stamer@gaos.org>
	* tpm_integrity.c: TPM_Quote2() added


2006-06-23  Mario Strasser <mast@gmx.net>
	* tpm_startup.c: behaviour of ST_CLEAR and storage of
		persistent data adapted


2006-06-18  Heiko Stamer <stamer@gaos.org>
	* tpm_identity.c: identity creation and activation,
		that is TPM_MakeIdentity and TPM_ActivateIdentity added


2005-12-24  Mario Strasser <mast@gmx.net>
	* tpm_transport.c, tpm_marshalling.c, tpm_structures.h:
		Transport session functionality added


2005-12-16  Mario Strasser <mast@gmx.net>
	* linux_module.c: ioctl() support added
	* tpm_capability.c: TPM_GetCapability() adapted to make
		it work with trousers


2005-12-09  Heiko Stamer <stamer@gaos.org>
	* tpm_marshalling.c, tpm_daa.c, rsa.c: DAA functionality,
		that is TPM_DAA_Join and TPM_DAA_Sign added


2005-08-15  Mario Strasser <mast@gmx.net>
	* all: some typos corrected
	* tpm_integrity.c: bug in TPM_Extend fixed


2005-05-07  Mario Strasser <mast@gmx.net>
	* bug in TPM_SaveContext fixed


2005-04-30  Mario Strasser <mast@gmx.net>
	* tpm_digest.c: audit functionality added
 

2005-04-29  Mario Strasser  <mast@gmx.net>
	* tpm_context.c: TPM_KeyControlOwner, TPM_SaveContext, 
		and TPM_LoadContext added
	* tpm_deprecated: TPM_(Save|Load)(Key|Auth)Context added
	* rc4.[ch]: RC4 encryption added
	* rsa.c: rsa_copy_key added


2005-04-27  Mario Strasser  <mast@gmx.net>
	* tpm_crypto.c: TPM_CertifyKey and TPM_CertifyKey2 added
	* tpm_deprecated: TPM_OwnerReadPubek, TPM_CertifySelfTest,
		TPM_DirWriteAuth, and TPM_DirRead added


2005-04-26  Mario Strasser  <mast@gmx.net>
	* tpm_ticks.c: timing ticks functionality added
	* tpm_authorization.c: TPM_ChangeAuth and TPM_ChangeAuthOwner added
	* tpm_storage, tpm_crypto.c, tpm_integrity.c: authorization 
		verification now considers the value of authDataUsage
	* tpm_storage: TPM_UnBind added


2004-10-24  Mario Strasser  <mast@gmx.net>
	
	* Makefile: new makefile for Linux kernels > 2.5.0 build system
	* tpm_testing.c: replacement of some floating-point calculations 
		with fix-point arithmetic 
	* all: minor changes int the directory and header file structure
	       	to make the package easier portable and maintainable 


2004-07-01  Mario Strasser  <mast@gmx.net>  

	* Initial release based on my semester thesis



crypto/sha1.c
/* Software-Based Trusted Platform Module (TPM) Emulator for Linux
 * Copyright (C) 2004 Mario Strasser <mast@gmx.net>,
 *
 * This module is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by the Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.
 *
 * This module is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * $Id$
 */

#include "sha1.h"

/* This code is based on Steve Reid's <steve@edmweb.com> 
   public domain implementation. */ 

#define rol(v,b) (((v) << (b)) | ((v) >> (32 - (b))))
#define B0(i) (buf[i] = CPU_TO_BE32(buf[i]))
#define B1(i) (buf[i & 15] = rol(buf[i & 15] ^ buf[(i-14) & 15] \
                                 ^ buf[(i-8) & 15] ^ buf[(i-3) & 15], 1))
#define F0(x,y,z) ((x & (y ^ z)) ^ z)
#define F1(x,y,z) (x ^ y ^ z)
#define F2(x,y,z) (((x | y) & z) | (x & y))
#define R0(a,b,c,d,e,i) e += F0(b,c,d) + B0(i) + 0x5A827999 + rol(a,5); b = rol(b,30);
#define R1(a,b,c,d,e,i) e += F0(b,c,d) + B1(i) + 0x5A827999 + rol(a,5); b = rol(b,30);
#define R2(a,b,c,d,e,i) e += F1(b,c,d) + B1(i) + 0x6ED9EBA1 + rol(a,5); b = rol(b,30);
#define R3(a,b,c,d,e,i) e += F2(b,c,d) + B1(i) + 0x8F1BBCDC + rol(a,5); b = rol(b,30);
#define R4(a,b,c,d,e,i) e += F1(b,c,d) + B1(i) + 0xCA62C1D6 + rol(a,5); b = rol(b,30);

static void sha1_transform(uint32_t h[5], const uint8_t data[64])
{
  uint32_t a, b, c, d, e;
  uint32_t buf[16];

  /* copy state and data*/
  a = h[0];
  b = h[1];
  c = h[2];
  d = h[3];
  e = h[4];
  memcpy(buf, data, 64);
  /* unrolled sha-1 rounds */
  R0(a,b,c,d,e, 0); R0(e,a,b,c,d, 1); R0(d,e,a,b,c, 2); R0(c,d,e,a,b, 3);
  R0(b,c,d,e,a, 4); R0(a,b,c,d,e, 5); R0(e,a,b,c,d, 6); R0(d,e,a,b,c, 7);
  R0(c,d,e,a,b, 8); R0(b,c,d,e,a, 9); R0(a,b,c,d,e,10); R0(e,a,b,c,d,11);
  R0(d,e,a,b,c,12); R0(c,d,e,a,b,13); R0(b,c,d,e,a,14); R0(a,b,c,d,e,15);
  R1(e,a,b,c,d,16); R1(d,e,a,b,c,17); R1(c,d,e,a,b,18); R1(b,c,d,e,a,19);
  R2(a,b,c,d,e,20); R2(e,a,b,c,d,21); R2(d,e,a,b,c,22); R2(c,d,e,a,b,23);
  R2(b,c,d,e,a,24); R2(a,b,c,d,e,25); R2(e,a,b,c,d,26); R2(d,e,a,b,c,27);
  R2(c,d,e,a,b,28); R2(b,c,d,e,a,29); R2(a,b,c,d,e,30); R2(e,a,b,c,d,31);
  R2(d,e,a,b,c,32); R2(c,d,e,a,b,33); R2(b,c,d,e,a,34); R2(a,b,c,d,e,35);
  R2(e,a,b,c,d,36); R2(d,e,a,b,c,37); R2(c,d,e,a,b,38); R2(b,c,d,e,a,39);
  R3(a,b,c,d,e,40); R3(e,a,b,c,d,41); R3(d,e,a,b,c,42); R3(c,d,e,a,b,43);
  R3(b,c,d,e,a,44); R3(a,b,c,d,e,45); R3(e,a,b,c,d,46); R3(d,e,a,b,c,47);
  R3(c,d,e,a,b,48); R3(b,c,d,e,a,49); R3(a,b,c,d,e,50); R3(e,a,b,c,d,51);
  R3(d,e,a,b,c,52); R3(c,d,e,a,b,53); R3(b,c,d,e,a,54); R3(a,b,c,d,e,55);
  R3(e,a,b,c,d,56); R3(d,e,a,b,c,57); R3(c,d,e,a,b,58); R3(b,c,d,e,a,59);
  R4(a,b,c,d,e,60); R4(e,a,b,c,d,61); R4(d,e,a,b,c,62); R4(c,d,e,a,b,63);
  R4(b,c,d,e,a,64); R4(a,b,c,d,e,65); R4(e,a,b,c,d,66); R4(d,e,a,b,c,67);
  R4(c,d,e,a,b,68); R4(b,c,d,e,a,69); R4(a,b,c,d,e,70); R4(e,a,b,c,d,71);
  R4(d,e,a,b,c,72); R4(c,d,e,a,b,73); R4(b,c,d,e,a,74); R4(a,b,c,d,e,75);
  R4(e,a,b,c,d,76); R4(d,e,a,b,c,77); R4(c,d,e,a,b,78); R4(b,c,d,e,a,79);
  /* update state */
  h[0] += a;
  h[1] += b;
  h[2] += c;
  h[3] += d;
  h[4] += e;
  /* overwrite all used variables */
  a = b = c = d = e = 0;
  memset(buf, 0, 64);
}


void sha1_init(sha1_ctx_t *ctx)
{
  /* initialise with sha-1 constants */
  ctx->h[0] = 0x67452301;
  ctx->h[1] = 0xEFCDAB89;
  ctx->h[2] = 0x98BADCFE;
  ctx->h[3] = 0x10325476;
  ctx->h[4] = 0xC3D2E1F0;
  ctx->count_lo = ctx->count_hi = 0;
}

void sha1_update(sha1_ctx_t *ctx, uint8_t *data, uint32_t length)
{
  uint32_t buf_off = (ctx->count_lo >> 3) & 63;
  uint32_t data_off = 0;

  /* add data */
  if (length + buf_off >= 64) {
    data_off = 64 - buf_off;
    memcpy(&ctx->buf[buf_off], data, data_off);
    sha1_transform(ctx->h, ctx->buf);
    while (data_off + 64 <= length) { 
      sha1_transform(ctx->h, &data[data_off]);
      data_off += 64;
    }
    buf_off = 0;
  }
  memcpy(&ctx->buf[buf_off], &data[data_off], length - data_off);
  /* update counter */
  buf_off = ctx->count_lo;
  ctx->count_lo += length << 3;
  if (ctx->count_lo < buf_off) ctx->count_hi++;
  ctx->count_hi += length >> 29;
}

void sha1_final(sha1_ctx_t *ctx, uint8_t digest[SHA1_DIGEST_LENGTH])
{
  uint8_t d, counter[8];

  /* setup counter */
  for (d = 0; d < 4; d++) {
    counter[d    ] = (ctx->count_hi >> (24 - d * 8)) & 0xff;
    counter[d + 4] = (ctx->count_lo >> (24 - d * 8)) & 0xff;
  }
  /* add padding */
  d = 0x80;
  sha1_update(ctx, &d, 1);
  d = 0x00;
  while ((ctx->count_lo & (63 * 8)) != (56 * 8)) sha1_update(ctx, &d, 1);
  /* add counter */
  sha1_update(ctx, counter, 8);
  for (d = 0; d < SHA1_DIGEST_LENGTH; d++) 
    digest[d] = (uint8_t)(ctx->h[d >> 2] >> (8 * (3 - (d & 3))) & 0xff);
  /* overwrite all used variables */
  memset(ctx, 0, sizeof(*ctx));
  memset(counter, 0, sizeof(counter));
}
